### PR TITLE
Complete implementation with support for fragments, introspection etc..

### PR DESCRIPTION
**Pull request by alexstrat**
This adds support for fragments, introspection #14 and more detailed errors reporting (and probably other features).

It is actually a revamp of the implementation to make it complete and make it closer to the [reference implementation](https://github.com/graphql/graphql-js/). Compared to the reference, this implementation uses the same structure, the same naming and uses directly some parts of the reference implementation when possible.

### Why revamp?

Initially, after trying to work on the support of fragments etc. by tweaking the current implementation, I figured out I was actually rewriting a large part of it, taking inspiration on the reference. I figured out it was easier to re-start from the reference directly.

Moreover, I also thought there are several benefits for the future to have an implementation closer to reference:
- easy to spot and backport fixes from the reference to `reactive-graphql`
- given the structure of the reference follows the specifications' structure, `reactive-graphql` implementation will follow the specifications' structure which makes it easier for a new-comer to understand
- easier to verify the reactive implementation is complete and spot the incomplete parts

_I understand it's not always easy for a maintainer to review and accept a revamp. That's why I'm making myself available to run you through it and to receive feedback, discussion, and change requests. 👐_

### Breaking changes
_for discussion_
#### Removed support for Observable `rootValue`
Was it really supported? Wow was it supported: root resolver got an Observable as `parent` parameter or emitted values one by one?

 _I think it can be easily supported if necessary_

####  Removed the support for Observable variables
Seems quite an odd feature.

If necessary, it  can be done by the user with a simple:
```js
variables$.pipe(
  switchMap(variables => graphql({ schema, source: query, variables }))
)
```

#### Removed support for query as `Document` in `graphql()`
Mapped to reference behaviour.
```js
// previously ok:
graphql(schema, gql`query {}`)

// now use:
execute(schema, gql`query {}`)
// or 
graphql(schema, `query {}`)
```
#### Changed error handling
Mapped to how `graphql-js` handles errors: every expected errors (validation, syntax error, throwing resolver...) are pushed in the `error` field of the `ExecutionResult`. 

### Todo
- [x] Fix Observable test
- [x] Explain the "implementation strategy"
- [x] Modify or remove some copy/pasted JSDoc
- [x] Move some parts "not reactive" from the `reactive-execute.ts` to `execute.ts`
- [x] Export `execute`
- [x] Add tests for `swtichMap` vs `concatMap`
- [x] Handle the left todos in code
- [x] Type with `unknown` instead of `any`
- [x] Modify Readme

### CODE REVIEW

**Comment by DanielMSchmidt** [comment]
Hey @alexstrat Thanks for this big initiative. Feel free to ping me when you are done and we can find some time in a chat to discuss it 👍

File: package.json [hunk 1]
```diff
@@ -19,14 +19,17 @@
     "build": "tsc -p tsconfig.dist.json"
   },
   "dependencies": {
-    "graphql": "^14.0.2"
+    "graphql": "^14.0.2",
```
Comment (alexstrat) [inline package.json]: Upgrade the version of `graphql`

**Comment by alexstrat** [comment]
The Mutation test case is failing after my implementation of `executeFieldsSerially` while I verified that, with a version of this test without using RxJS Marbles, the test is passing.

My guess is that [the Promise (`async`)-based trick I used to implement seriality](https://github.com/mesosphere/reactive-graphql/pull/28/files#diff-18c4bef481c0f311f14a1dd513f0dc09R247-R273) is not accepted by Marbles testing because [RxJS test code that consumes Promise is not possible](https://rxjs-dev.firebaseapp.com/guide/testing/marble-testing#you-can-t-directly-test-rxjs-code-that-consumes-promises-or-uses-any-of-the-other-schedulers-e-g-asapscheduler-).

I need to review this trick to use RxJs magic only. Any idea is welcome!

**Comment by alexstrat** [comment]
@DanielMSchmidt I think I'm done and available for a chat

**Comment by alexstrat** [comment]
@DanielMSchmidt do you think you could publish an `-rc` version with this PR?

**Comment by DanielMSchmidt** [comment]
@alexstrat Sorry, forgot to post my last response

>  I think I'm done and available for a chat when you are.

I try to go through the entire PR until friday, maybe we can chat on friday and clear up things that are not clear to me yet :) As the implementation changes I think we should provide a migration guide / codemod.

> do you think you could publish an -rc version with this PR?

Did so, we have a 4.0.0-rc.1 

That being said, we should squash your commits and add a `BREAKING CHANGE: ` in the commit body, so that semantic release can deploy the 4.0.0 version properly

**Comment by alexstrat** [comment]
@DanielMSchmidt thanks for the RC!

I'm sending you an email on your gmail address to find the time to chat!

Regarding the migration guide / codemod: yes good idea! That's what I started to do in the _Breaking changes_ part of this PR's description: in comparison, what kind of format and location for a migration guide you had in mind?

As for the codemod, given `reactive-graphql` is still (as far as I can guess, of course) a narrowly-used project, I'm not sure the benefit outweighs the effort required to make a codemod. Though, I haven't worked on any codemod so far, I'm maybe overestimating this effort.

Regarding the squash to make it semantic release friendly, you want me to rewrite the history of the PR and make a squash or you can do it with the merge button?

File: src/rxutils/mapPromiseToObservale.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import { Observable, from } from "rxjs";
+import { switchMap } from "rxjs/operators";
+
+/**
+ * Build an Observable given a Promise's resolved value.
+ *
+ * @param promise 
+ * @param observableFn 
+ */
+export default function mapPromiseToObservale<TPromiseValue, TObservableValue>(
```
Comment (alexstrat) [inline src/rxutils/mapPromiseToObservale.ts]: Minor typo

File: package.json [hunk 3]
```diff
@@ -19,16 +19,19 @@
     "build": "tsc -p tsconfig.dist.json"
   },
   "dependencies": {
-    "graphql": "^14.0.2"
```
Comment (alexstrat) [inline package.json]: I don't think it's ok

**Pull request merged**